File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    //brute force O(n^2)
//     let profit =0
    
//     for(let i=0;i<prices.length-1;i++){
//         for(let j=i+1;j<prices.length ;j++){
//             const currProfit=prices[j]-prices[i]
//             profit=Math.max(profit,currProfit)
//         }
//     }
//     return profit
    
    let profit=0
    
    let buy=0
    let sell=1
    while(sell<prices.length){
        
        if(prices[buy]<prices[sell]){
            const currProfit=prices[sell]-prices[buy]
        profit=Math.max(profit,currProfit)
            
        }
        else buy=sell
        sell++
    }
    return profit
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # profit=0
        # buy=0
        # sell=1
        # while sell<len(prices):
        #     if prices[buy]<prices[sell]:
        #         curr_profit=prices[sell]-prices[buy]
        #         profit=max(profit,curr_profit)
        #     else:
        #         buy=sell
        #     sell+=1
        # return profit
        if not prices:
            return 0

        min_price = prices[0]
        max_profit = 0

        for price in prices:
            min_price = min(min_price, price)
            profit = price - min_price
            max_profit = max(max_profit, profit)

        return max_profit
        
        
        
        
        
        

        